### ISSUE TITLE
PW3dev Uploading images problem

### ISSUE THREAD

**Issue by tbba** [ryancramerdesign/ProcessWire#1557]
Unfortunately I found a case where the image upload problem does still happen here
(localhost/MAMP *and* live; reno and standard theme, PW 3.02dev - does not happen with PW 2):

It happens with both image crop modules, now tested with croppable:
I upload an image by dragging it to the croppable field.
- The upload indicator shows 100%
- The file was actually uploaded (I checked that with FTP)
- The image is not shown in the field and after page save, the field appears to be empty. 
- No error msg, just a stalling (spinner image rotates forever)

The field was setup like this: 
- croppableImage
- max allowed: 1
- formatting: one element or NULL
- standard view: row 
- min+max set 
- refuse to large image: set

maybe related:
https://processwire.com/talk/topic/8712-imagefile-upload-problems-without-errors/
(Maybe there is a solution without the need of the suggested tweaking of Apache - because it *HAD* worked)

**Comment by teppokoivula** [comment]
Could you check what the response of the AJAX request contains? This can be checked via dev tools of your browser, and could provide additional details about what really happened behind the scenes.

Am I reading this right, i.e. this issue only occurs with 3rd party crop modules installed/enabled?

**Comment by adrianbj** [comment]
The problem (or something related) also occurs with ImageExtra. In that case the problem was a property of non object error pointing to this line:
```
    $p = $event->object->value['page'];
```
I didn't actually need the module and am supposed to be on vacation at the moment, so didn't investigate at all, but thought I'd share here in case it's helpful.

**Comment by ryancramerdesign** [comment]
I will give ImageExtra a try to see what the issue might be, just in case it's FileCompiler or PW3 hooks related. Though I've noticed when namespaces are active PHP is a little more strict on some things, throwing notices when it doesn't without namespaces, and this might be one of this instances.

**Comment by tbba** [comment]
Every attempt to fix this would be welcome. Right now this is a reasons why I had to downgrade to be able to continue working on the project. (in 2.x everything works, the trouble came with 3.x)
That is why I cannot do further testings about AJAX requests right now. I have downgraded already.

The trouble in 2.x is that If I include an image in an (inline 2languages) CKE field that placement is lost on save. If I click in another field before saving (I mean: blurring the field) I can save the page and the changes to the cke field are kept. - This might be a Safari related issue. I am traveling right now and only have Safari at the moment.

**Comment by teppokoivula** [comment]
Checked the AJAX request with CroppableImage installed myself, and this is what I'm seeing:

> Recoverable Fatal Error: 	Argument 1 passed to InputfieldCroppableImage::getAdminThumbCroppable() must be an instance of PageImage, instance of ProcessWire\Pageimage given, called in /site/assets/cache/FileCompiler/site/modules/CroppableImage/InputfieldCroppableImage/InputfieldCroppableImage.module on line 119 and defined (line 60 of /site/assets/cache/FileCompiler/site/modules/CroppableImage/InputfieldCroppableImage/InputfieldCroppableImage.module) 
outpori.fi/site/assets/cache/FileCompiler/site/modules/CroppableImage/InputfieldCroppableImage/InputfieldCroppableImage.module)

**Comment by teppokoivula** [comment]
I'm pretty sure this was discussed elsewhere too, but the issue seems to be related to case-sensitivity. Following change to FileCompiler.php fixes this, but could result in performance loss, and @ryancramerdesign might have a better solution :)

```
diff --git a/wire/core/FileCompiler.php b/wire/core/FileCompiler.php
index 3dc0b7f..67d65d6 100644
--- a/wire/core/FileCompiler.php
+++ b/wire/core/FileCompiler.php
@@ -402,11 +402,11 @@ class FileCompiler extends Wire {
                                "instanceof" => '(\sinstanceof\s+)' . $class . '([^_a-zA-Z0-9]|$)', // 'instanceof Page'
                                "$class " => '(\(\s*|,\s*)' . $class . '(\s+\$)', // type hinted '(Page $something' or '($foo, Page $something'
                        );

                        foreach($patterns as $check => $regex) {

-                               if(strpos($data, $check) === false) continue;
-                               if(!preg_match_all('/' . $regex . '/m', $data, $matches)) continue;
+                               if(stripos($data, $check) === false) continue;
+                               if(!preg_match_all('/' . $regex . '/mi', $data, $matches)) continue;
                                //echo "<pre>" . print_r($matches, true) . "</pre>";

                                foreach($matches[0] as $key => $fullMatch) {
```

**Comment by ryancramerdesign** [comment]
If it's a case sensitivity issue, I actually just completed some updates to the FileCompiler earlier this morning that supports situations where a module or template uses the improper case when mentioning core classes, among some other tweaks. For the case sensitivity part, I think it's basically exactly what @teppokoivula mentioned above. 3rd party modules really should use the correct case when referring to core classes and such (which appears to be the source of that error in CroppableImage and ImageExtra), but since PHP apparently doesn't require it, FileCompiler doesn't need to either.

**Comment by teppokoivula** [comment]
Confirmed. Latest version of 3.x does indeed fix this issue for me, at least in case of CroppableImage.

**Comment by tbba** [comment]
HOORAY!! FIXED!!! - Now I can switch back to 3.0.
